vcd/resource_vcd_independent_disk.go

```diff
@@ -173,33 +197,180 @@ func resourceVcdIndependentDiskCreate(d *schema.ResourceData, meta interface{})
 
 	diskCreateParams.Disk.Description = d.Get("description").(string)
 
+	if value, ok := d.GetOk("sharing_type"); ok {
+		diskCreateParams.Disk.SharingType = value.(string)
+	}
+
 	task, err := vdc.CreateDisk(diskCreateParams)
 	if err != nil {
-		return fmt.Errorf("error creating independent disk: %s", err)
+		return diag.Errorf("error creating independent disk: %s", err)
 	}
 
 	err = task.WaitTaskCompletion()
 	if err != nil {
-		return fmt.Errorf("error waiting to finish creation of independent disk: %s", err)
+		return diag.Errorf("error waiting to finish creation of independent disk: %s", err)
 	}
 
 	diskHref := task.Task.Owner.HREF
 	disk, err := vdc.GetDiskByHref(diskHref)
 	if err != nil {
-		return fmt.Errorf("unable to find disk with href %s: %s", diskHref, err)
+		return diag.Errorf("unable to find disk with href %s: %s", diskHref, err)
 	}
 
 	d.SetId(disk.Disk.Id)
 
-	return resourceVcdIndependentDiskRead(d, meta)
+	return resourceVcdIndependentDiskRead(ctx, d, meta)
+}
+
+func resourceVcdIndependentDiskUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	if d.HasChanges("size_in_mb", "storage_profile", "description") {
+		_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+		if err != nil {
+			return diag.Errorf(errorRetrievingOrgAndVdc, err)
+		}
+
+		storageProfileValue := d.Get("storage_profile").(string)
+		var storageProfileRef *types.Reference
+
+		if storageProfileValue != "" {
+			storageReference, err := vdc.FindStorageProfileReference(storageProfileValue)
+			if err != nil {
+				return diag.Errorf("error finding storage profile %s", storageProfileValue)
+			}
+			storageProfileRef = &types.Reference{HREF: storageReference.HREF}
+		}
+
+		disk, err := vdc.GetDiskById(d.Id(), true)
+		if err != nil {
+			return diag.Errorf("error fetching independent disk: %s", err)
+		}
+
+		sliceOfVmsHrefs, err := disk.GetAttachedVmsHrefs()
+		if err != nil {
+			return diag.Errorf("[ERROR] resourceVcdIndependentDiskUpdate faced issue fetching attached VMs")
+		}
+
+		//lock VMs as another independent disk resource can be doing update with same VM
+		lockVms(sliceOfVmsHrefs)
```

Comment:
Can we get into a deadlock here?

This is not an atomic lock and while looping over VMs, a few VMs can have another disk whose update operation has already locked one of the VMs. As a result both of these "lock loops" could be in deadlock.

Reply:
The possibility exists for shared disks. As only sliceOfVmsHrefs will be >1 if the disk is shared. In that case, if there are two disks attached to 2 VMs and both of them are shared and they want to acquire VM1 and VM2 lock. 
I couldn't make this "atomic" as let's say disk1 attached to VM1 and disk2 attached to VM1 and VM2 - this means locks are different and resource will fail as they will be changing VM1 at the same time. So deadlock is a corner case for every other cases. If anyone sees a better approach, please share.

Reply:
After discussion with Dainius we agreed to have an additional global lock for shared independent disks, this allows avoiding dead locks. Yet make them a little slower on updates.